src/crimson/os/seastore/seastore.cc

```diff
@@ -984,16 +984,16 @@ seastar::future<> SeaStore::do_transaction(
     op_type_t::TRANSACTION,
     [this](auto &ctx) {
       return with_trans_intr(*ctx.transaction, [&, this](auto &t) {
-        return onode_manager->get_or_create_onodes(
-          *ctx.transaction, ctx.iter.get_objects()
-        ).si_then([this, &ctx](auto &&onodes) {
-          return seastar::do_with(std::move(onodes), [this, &ctx](auto& onodes) {
-            return trans_intr::repeat(
-              [this, &ctx, &onodes]() -> tm_iertr::future<seastar::stop_iteration>
+        return seastar::do_with(std::vector<OnodeRef>(ctx.iter.objects.size()),
+          std::vector<OnodeRef>(),
+          [this, &ctx](auto& onodes, auto& d_onodes) mutable {
+          return trans_intr::repeat(
+            [this, &ctx, &onodes, &d_onodes]() mutable
+            -> tm_iertr::future<seastar::stop_iteration>
```

Comment:
I think we need to clear onodes and d_onodes after each repeat in case of transaction conflicts.

Reply:
onodes and d_onodes shouldn't be clear here.  repeat is for next op, if ctx.iter.has_op() is true, the repeat will continue, each time do_transaction_step only deal one op in ctx.iter, and the next op maybe use last time created onodeRef, we only need guarantee onode and d_onode will not have duplicated OnodeRef, so before get_onode or create onode will check if onode[op->oid] exist,  and the same, before set onode[op->oid] = get_onode or push to d_onodes, check if it already exists in onodes.  So I think when transaction conflicts and we repeat here, if onodeRef already exists, will not get or create onodeRef again.

Reply:
a little mistake, repeat_eagain happen in repeat_with_internal_context, at that situation,  all nodes and d_onodes in do_with will execute again and cleared. trans_intr::repeat is not for transaction confliction, it means this repeat loop is interruptible.